### PR TITLE
Add regression tests for a cloud-based Neon instance

### PR DESCRIPTION
**Pull request by a-masterov**
## Problem
We need to be able to run the regression tests against a cloud-based Neon instance in order to prepare the migration to the arm architecture.

## Summary of changes
Some tests were modified to work on the cloud instance (i.e. added passwords, server-side copy changed to client-side, etc)

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: .github/workflows/cloud-regress.yml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    #          │ │ ┌───────────── day of the month (1 - 31)
+    #          │ │ │ ┌───────────── month (1 - 12 or JAN-DEC)
+    #          │ │ │ │ ┌───────────── day of the week (0 - 6 or SUN-SAT)
+    - cron:   '0 3 * * *' # run once a day, timezone is utc
+  workflow_dispatch: # adds ability to run this manually
+    inputs:
+      region_id:
+        description: 'Project region id. If not set, the default region will be used'
+        required: false
+        default: 'aws-us-east-2'
```
[the remark this answers is not in the document]
Comment (a-masterov) [reply .github/workflows/cloud-regress.yml:19]: Removed.

File: patches/cloud_regress_pg16.patch [hunk 2]
```diff
@@ -0,0 +1,3949 @@
...
+--- a/src/test/regress/expected/alter_generic.out
++++ b/src/test/regress/expected/alter_generic.out
+@@ -15,9 +15,9 @@ DROP ROLE IF EXISTS regress_alter_generic_user1;
+ DROP ROLE IF EXISTS regress_alter_generic_user2;
+ DROP ROLE IF EXISTS regress_alter_generic_user3;
+ RESET client_min_messages;
+-CREATE USER regress_alter_generic_user3;
+-CREATE USER regress_alter_generic_user2;
+-CREATE USER regress_alter_generic_user1 IN ROLE regress_alter_generic_user3;
++CREATE USER regress_alter_generic_user3 PASSWORD 'eib5tai9Oachootiex';
```
[the remark this answers is not in the document]
Comment (a-masterov) [reply patches/cloud_regress_pg16.patch:26]: Unfortunately, no, if I am not mistaken
```
ERROR:  Received HTTP code 400 from control plane: {"error":"Neon only supports being given plaintext passwords"}
```

File: .github/workflows/cloud-regress.yml [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      options: --init
+
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          submodules: true
+
+      - name: Patch the test
+        run:
+          cd /__w/neon/neon/vendor/postgres-v${{ matrix.DEFAULT_PG_VERSION }} && ls && patch -p1 < ../../patches/cloud_regress_pg${{ matrix.DEFAULT_PG_VERSION }}.patch
```
[the remark this answers is not in the document]
Comment (a-masterov) [reply .github/workflows/cloud-regress.yml]: Fixed, thanks

File: test_runner/cloud_regress/test_cloud_regress.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+@pytest.mark.remote_cluster
+def test_cloud_regress(remote_pg: RemotePostgres):
+    """
+    Run the regression tests
+    """
+    cur_test = os.environ.get("PYTEST_CURRENT_TEST")
+    assert cur_test
+    pg_version_match = re.search(r"\-pg(\d+)\]", cur_test)
+    assert pg_version_match
+    pg_version = int(pg_version_match.group(1))
```
[the remark this answers is not in the document]
Comment (a-masterov) [reply test_runner/cloud_regress/test_cloud_regress.py:25]: Good catch, thanks

File: test_runner/cloud_regress/test_cloud_regress.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            else:
+                log.info("Creating the extension")
+                cur.execute("CREATE EXTENSION regress_so")
+                conn.commit()
+
+            log.info("Creating a C function to check availability of regress.so")
+            cur.execute(
+                "CREATE FUNCTION get_columns_length(oid[]) "
+                "RETURNS int AS 'regress.so' LANGUAGE C STRICT STABLE PARALLEL SAFE;"
+            )
```
[the remark this answers is not in the document]
Comment (a-masterov) [reply test_runner/cloud_regress/test_cloud_regress.py:36]: This is also a workaround for the full path problem

File: test_runner/cloud_regress/test_cloud_regress.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+                "--max-connections=5",
+            ]
+            try:
+                remote_pg.pg_bin.run(regress_cmd, env=env_vars, cwd=runpath)
+            except subprocess.CalledProcessError as e:
+                log.error("Error(s) occurred while running the regression tests")
+                with open(f"{runpath}/regression.out", "r") as f:
+                    print(f.read())
+                with open(f"{runpath}/regression.diffs", "r") as f:
+                    print(f.read())
```
[the remark this answers is not in the document]
Comment (a-masterov) [reply test_runner/cloud_regress/test_cloud_regress.py:72]: Fixed

File: .github/workflows/cloud-regress.yml [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        default: 'aws-us-east-2'
+
+defaults:
+  run:
+    shell: bash -euxo pipefail {0}
+
+concurrency:
+  # Allow only one workflow per any non-`main` branch.
+  group: ${{ github.workflow }}-${{ github.ref_name }}-${{ github.ref_name == 'main' && github.sha || 'anysha' }}
+  cancel-in-progress: true
```
[the remark this answers is not in the document]
Comment (a-masterov) [reply .github/workflows/cloud-regress.yml]: Done, thanks

File: test_runner/cloud_regress/test_cloud_regress.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+                "CREATE FUNCTION get_columns_length(oid[]) "
+                "RETURNS int AS 'regress.so' LANGUAGE C STRICT STABLE PARALLEL SAFE;"
+            )
+            conn.rollback()
+            runpath = (
+                f"{os.path.abspath(f'{os.path.relpath(__file__)}/../../../')}"
+                f"/vendor/postgres-v{pg_version}/src/test/regress"
+            )
+            prefix = f"/tmp/neon/pg_install/v{pg_version}"
+            regress_bin = f"{prefix}/lib/postgresql/pgxs/src/test/regress/pg_regress"
```
[the remark this answers is not in the document]
Comment (a-masterov) [reply test_runner/cloud_regress/test_cloud_regress.py]: fixed

File: .github/workflows/cloud-regress.yml [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        with:
+          build_type: ${{ env.BUILD_TYPE }}
+          test_selection: cloud_regress
+          run_in_parallel: false
+          save_perf_report: ${{ env.SAVE_PERF_REPORT }}
+          pg_version: ${{ env.DEFAULT_PG_VERSION }}
+          extra_params:
+            -m remote_cluster
+        env:
+          BENCHMARK_CONNSTR: ${{ secrets.PG_REGRESS_CONNSTR }}
```
[the remark this answers is not in the document]
Comment (a-masterov) [reply .github/workflows/cloud-regress.yml]: done

File: test_runner/cloud_regress/test_cloud_regress.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bayandin) [reply test_runner/cloud_regress/test_cloud_regress.py:37]: Should we fix the problem with the full path? We know the approach, right?

File: .github/workflows/cloud-regress.yml [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+  group: ${{ github.workflow }}
+  cancel-in-progress: true
+
+jobs:
+  regress:
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - DEFAULT_PG_VERSION: 16
```
[the remark this answers is not in the document]
Comment (a-masterov) [reply .github/workflows/cloud-regress.yml]: Fixed

File: .github/workflows/cloud-regress.yml [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+        run: |
+          cd /__w/neon/neon/vendor/postgres-v${{ matrix.DEFAULT_PG_VERSION }}
+          patch -p1 < ../../patches/cloud_regress_pg${{ matrix.DEFAULT_PG_VERSION }}.patch
+
+      - name: Generate a random password
+        id: pwgen
+        run: |
+          set +x
+          DBPASS=$(dd if=/dev/random bs=48 count=1 2>/dev/null | base64)
+          echo DBPASS="${DBPASS//\//}" >> "${GITHUB_OUTPUT}"
```
[the remark this answers is not in the document]
Comment (a-masterov) [reply .github/workflows/cloud-regress.yml:55]: Done

File: test_runner/cloud_regress/test_cloud_regress.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (a-masterov) [reply test_runner/cloud_regress/test_cloud_regress.py:54]: If I understand correctly, this problem depends on the environment variables passing through the proxy problem.
PG_LIBDIR is not set due to this.

File: test_runner/cloud_regress/test_cloud_regress.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+            cur.execute(
+                "SELECT rolname FROM pg_catalog.pg_roles WHERE oid > 16384 AND rolname <> 'neondb_owner'"
+            )
+            roles: list[Any] = []
+            for role in cur:
+                log.info("Role found: %s", role[0])
+                roles.append(role[0])
+            for role in roles:
+                cur.execute(f"DROP ROLE {role}")
+            conn.commit()
```
Comment (bayandin) [inline test_runner/cloud_regress/test_cloud_regress.py:69]: I wonder if we should create a branch and run tests on the Neon branch instead of doing this cleanup?
This also will allow us to run several workflow runs in parallel

File: .github/workflows/cloud-regress.yml [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+    #          │ │ │ │ ┌───────────── day of the week (0 - 6 or SUN-SAT)
+    - cron:   '0 3 * * *' # run once a day, timezone is utc
+  workflow_dispatch: # adds ability to run this manually
+
+defaults:
+  run:
+    shell: bash -euxo pipefail {0}
+
+concurrency:
+  # Allow only one workflow per any non-`main` branch.
```
Comment (bayandin) [inline .github/workflows/cloud-regress.yml:20]: This comment is not relevant

File: .github/workflows/cloud-regress.yml [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+        if: ${{ !cancelled() }}
+        uses: ./.github/actions/allure-report-generate
+
+      - name: Post to a Slack channel
+        if: ${{ github.event.schedule && failure() }}
+        uses: slackapi/slack-github-action@v1
+        with:
+          channel-id: "C033QLM5P7D" # on-call-staging-stream
+          slack-message: |
+            Periodic regression testing: ${{ job.status }}
```
[the remark this answers is not in the document]
Comment (a-masterov) [reply .github/workflows/cloud-regress.yml]: done

File: test_runner/cloud_regress/test_cloud_regress.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lubennikovaav) [reply test_runner/cloud_regress/test_cloud_regress.py:54]: The problem is that pg_regress test setup sets libdir part of the path explicitly
https://github.com/neondatabase/postgres/blob/REL_16_STABLE_neon/src/test/regress/sql/test_setup.sql#L10

instead of using `$libdir` macro as all other extensions do
https://github.com/neondatabase/postgres/blob/3ec6e2496f64c6fec35c67cb82efd6490a6a4738/src/pl/plpgsql/src/plpgsql.control#L4

We can either fix handling of this in our extension request code (this will help other extensions that use full path as well)
or  patch `/test_setup.sql` to make it use $libdir

```
diff --git a/src/test/regress/sql/test_setup.sql b/src/test/regress/sql/test_setup.sql
index 1b2d434683..20c45b903c 100644
--- a/src/test/regress/sql/test_setup.sql
+++ b/src/test/regress/sql/test_setup.sql
@@ -7,7 +7,7 @@
 \getenv libdir PG_LIBDIR
 \getenv dlsuffix PG_DLSUFFIX
 
-\set regresslib :libdir '/regress' :dlsuffix
+\set regresslib '$libdir/regress' :dlsuffix
 
 --
 -- synchronous_commit=off delays when hint bits may be set. Some plans change
```

File: test_runner/cloud_regress/test_cloud_regress.py (hunk 12 above)
Comment (bayandin) [reply test_runner/cloud_regress/test_cloud_regress.py:69]: It think it's a nice think to do eventually (using branching), but let's do in a separate PR

File: test_runner/cloud_regress/test_cloud_regress.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (a-masterov) [reply test_runner/cloud_regress/test_cloud_regress.py:54]: This is a good solution, but it's a workaround.
I believe https://github.com/neondatabase/postgres/pull/500 is a true solution for this problem.

File: .github/workflows/cloud-regress.yml [hunk 15]
```diff
@@ -0,0 +1,104 @@
+name: Cloud Regression Test
+on:
+  push:
+    branches: [amasterov/regress-arm]
```
Comment (lubennikovaav) [inline .github/workflows/cloud-regress.yml:4]: I don't know how to write workflow yamls, but I guess this must be changed to mention `main` / `release` branch before merge?

File: .github/workflows/cloud-regress.yml [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+      POSTGRES_DISTRIB_DIR: /tmp/neon/pg_install
+      DEFAULT_PG_VERSION: 16
+      TEST_OUTPUT: /tmp/test_output
+      BUILD_TYPE: remote
+      AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_DEV }}
+      AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_KEY_DEV }}
+
+    runs-on: us-east-2
+    container:
+      image: neondatabase/build-tools:pinned
```
Comment (lubennikovaav) [inline .github/workflows/cloud-regress.yml:36]: Why do we use `pinned` image here?
please, add a comment

File: .github/workflows/cloud-regress.yml [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+
+      - name: Run the regression tests
+        uses: ./.github/actions/run-python-test-set
+        with:
+          build_type: ${{ env.BUILD_TYPE }}
+          test_selection: cloud_regress
+          pg_version: ${{ env.DEFAULT_PG_VERSION }}
+          extra_params: -m remote_cluster
+        env:
+          BENCHMARK_CONNSTR: ${{ secrets.PG_REGRESS_CONNSTR }}
```
Comment (lubennikovaav) [inline .github/workflows/cloud-regress.yml:86]: Q: does this test reuses the same database as benchmark or we generate new project for each test run?

File: patches/cloud_regress_pg16.patch [hunk 18]
```diff
@@ -0,0 +1,3949 @@
...
+index 0c24f6afe4..dd808ac2b4 100644
+--- a/src/test/regress/expected/aggregates.out
++++ b/src/test/regress/expected/aggregates.out
+@@ -11,7 +11,8 @@ CREATE TABLE aggtest (
+ 	b			float4
+ );
+ \set filename :abs_srcdir '/data/agg.data'
+-COPY aggtest FROM :'filename';
++\set command '\\copy aggtest FROM ' :'filename';
++:command
```
Comment (lubennikovaav) [inline patches/cloud_regress_pg16.patch:11]: Q: Why this change was needed?

File: test_runner/cloud_regress/test_cloud_regress.py [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+                            regress_cur.execute(f"ALTER SUBSCRIPTION {sub[0]} DISABLE")
+                            regress_cur.execute(
+                                f"ALTER SUBSCRIPTION {sub[0]} SET (slot_name = NONE)"
+                            )
+                            regress_cur.execute(f"DROP SUBSCRIPTION {sub[0]}")
+                            regress_conn.commit()
+
+            # This is also a workaround for the full path problem
+            # If we specify the full path in the command, the library won't be downloaded
+            # So we specify the name only for the first time
```
Comment (lubennikovaav) [inline test_runner/cloud_regress/test_cloud_regress.py:49]: stale comment?
```suggestion
```

File: test_runner/cloud_regress/test_cloud_regress.py [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+    with psycopg2.connect(remote_pg.connstr()) as conn:
+        with conn.cursor() as cur:
+            cur = conn.cursor()
+            log.info("Creating the extension")
+            cur.execute("CREATE EXTENSION IF NOT EXISTS regress_so")
+            conn.commit()
+            log.info("Looking for subscriptions in the regress database")
+            cur.execute(
+                "SELECT subname FROM pg_catalog.pg_subscription WHERE "
+                "subdbid = (SELECT oid FROM pg_catalog.pg_database WHERE datname='regression');"
```
Comment (lubennikovaav) [inline test_runner/cloud_regress/test_cloud_regress.py:29]: Why would regression database have subscriptions?

File: test_runner/cloud_regress/test_cloud_regress.py [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+        with conn.cursor() as cur:
+            cur.execute(
+                "SELECT rolname FROM pg_catalog.pg_roles WHERE oid > 16384 AND rolname <> 'neondb_owner'"
+            )
+            roles: list[Any] = []
+            for role in cur:
+                log.info("Role found: %s", role[0])
+                roles.append(role[0])
+            for role in roles:
+                cur.execute(f"DROP ROLE {role}")
```
Comment (lubennikovaav) [inline test_runner/cloud_regress/test_cloud_regress.py:68]: I assume that we run tests in an empty clear project, but it seems not to be the case...

Please, add a comment, why this is necessary and TODO for the future plans

**Review by lubennikovaav — CHANGES_REQUESTED** [review]

File: .github/workflows/cloud-regress.yml (hunk 15 above)
Comment (a-masterov) [reply .github/workflows/cloud-regress.yml:4]: I will remove this before merging, thanks

File: .github/workflows/cloud-regress.yml (hunk 17 above)
Comment (a-masterov) [reply .github/workflows/cloud-regress.yml:86]: This test re-uses the same project for now. 
It's required to grant the superuser privileges and I don't know how to automate this process.

File: patches/cloud_regress_pg16.patch (hunk 18 above)
Comment (a-masterov) [reply patches/cloud_regress_pg16.patch:11]: We are running the regression test against the cloud instance (staging).
This instance does not have the files, which are available for pg_regress. That's why we changed the server-side copy to the client-side.

File: test_runner/cloud_regress/test_cloud_regress.py (hunk 19 above)
Comment (a-masterov) [reply test_runner/cloud_regress/test_cloud_regress.py]: Thank you, the comments are now removed with the workaround.

File: test_runner/cloud_regress/test_cloud_regress.py (hunk 21 above)
Comment (a-masterov) [reply test_runner/cloud_regress/test_cloud_regress.py:61]: I have added a TODO comment. Thanks!

File: .github/workflows/cloud-regress.yml (hunk 16 above)
Comment (a-masterov) [reply .github/workflows/cloud-regress.yml:36]: We use the `pinned` build-tools image almost everywhere. I believe to avoid storage over-consumption and speed up building.
please, refer to https://github.com/neondatabase/neon/blob/amasterov/regress-arm/.github/workflows/pin-build-tools-image.yml

**Review by lubennikovaav — APPROVED** [review]

**Review by bayandin — APPROVED** [review]

File: .github/workflows/cloud-regress.yml (hunk 15 above)
Comment (a-masterov) [reply .github/workflows/cloud-regress.yml]: Removed

File: .github/workflows/cloud-regress.yml [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+        run: |
+          cd vendor/postgres-v"${DEFAULT_PG_VERSION}"/src/test/regress
+          for fname in sql/*.sql expected/*.out; do
+            sed -i.bak s/NEON_PASSWORD_PLACEHOLDER/"'${DBPASS}'"/ "${fname}"
+          done
+          for ph in $(grep NEON_MD5_PLACEHOLDER expected/password.out | awk '{print $3;}' | sort | uniq); do
+            USER=$(echo "${ph}" | cut -c 22-)
+            MD5=md5$(echo -n "${DBPASS}${USER}" | md5sum | awk '{print $1;}')
+            sed -i.bak "s/${ph}/${MD5}/" expected/password.out
+          done
```
[the remark this answers is not in the document]
Comment (a-masterov) [reply .github/workflows/cloud-regress.yml:67]: Dear Rabbit, we need to test the MD5 password hashing implemented in Postgres, that's why we use this algorithm. I know that it is not considered secure anymore, but don't worry, the accounts will exist at most several seconds so, it cannot be considered as a security vulnerability.
Please feel free to speak up with any suggestions or concerns.
Regards.

File: test_runner/cloud_regress/test_cloud_regress.py [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+
+
+@pytest.fixture
+def setup(remote_pg: RemotePostgres):
+    """
+    Setup and teardown of the tests
+    """
+    with psycopg2.connect(remote_pg.connstr()) as conn:
+        with conn.cursor() as cur:
+            cur = conn.cursor()
```
[the remark this answers is not in the document]
Comment (a-masterov) [reply test_runner/cloud_regress/test_cloud_regress.py]: Fixed, thanks

File: test_runner/cloud_regress/test_cloud_regress.py [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+                ) as regress_conn:
+                    with regress_conn.cursor() as regress_cur:
+                        for sub in cur:
+                            regress_cur.execute(f"ALTER SUBSCRIPTION {sub[0]} DISABLE")
+                            regress_cur.execute(
+                                f"ALTER SUBSCRIPTION {sub[0]} SET (slot_name = NONE)"
+                            )
+                            regress_cur.execute(f"DROP SUBSCRIPTION {sub[0]}")
+                            regress_conn.commit()
+
```
[the remark this answers is not in the document]
Comment (a-masterov) [reply test_runner/cloud_regress/test_cloud_regress.py]: Fixed, thanks
